### PR TITLE
Relaxing ontology label matching (SCP-3370)

### PR DESCRIPTION
**Pull request by bistline**
Currently, ingest pipeline enforces strict matching on all required ontology label columns in convention metadata files.  While this ensures a high degree of uniformity, it has the large downside of generating largely pointless errors when OLS updates an ontology label (such as recent changes to "10x" sequencing technology labels) and a user has not used the updated term in their metadata file.

This update relaxes this requirement to allow both case-insensitive matches, and any approved synonyms or related synonyms for a given ontology entry.  For example:

A user provides `10x 3' v2 sequencing` as the value for `library_preparation_protocol__ontology_label`, using the ontology ID of `EFO_0009899`.  The actual label according to the [EFO ontology](https://www.ebi.ac.uk/ols/ontologies/efo/terms?iri=http%3A%2F%2Fwww.ebi.ac.uk%2Fefo%2FEFO_0009899) is `10x 3' v2`, but `10x 3' v2 sequencing` is listed as an official synonym, as is `10X 3' v2`.  This value would be marked acceptable under these new conditions, whereas previously it would have thrown an error and caused the file to fail validation, and be deleted from the user's study.

TO TEST:
1. Pull this branch and run the installation instructions from the [README](https://github.com/broadinstitute/scp-ingest-pipeline#install)
2. Navigate to `ingest/validation`
3. Run `python3 validate_metadata.py ../../tests/data/annotation/metadata/convention/valid_no_array_synonyms_v2.0.0.txt`
4. Note there are no errors, only warnings
5. Validate script returned ok by checking `echo $?`, which will return `0`

This PR satisfies SCP-3370.

### CODE REVIEW

**Review by eweitz — CHANGES_REQUESTED** [review]
This will be a nice user quality-of-life improvement -- and for us implementers, too.  No more scrambling due to such trivial upstream changes!

With a few refinements to prevent maintainers from stumbling, I think this will be good to go.

File: ingest/validation/validate_metadata.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ class OntologyRetriever:
     cached_ontologies = {}
     cached_terms = {}
 
-    def retrieve_ontology_term_label(self, term, property_name, convention, attribute_type: str):
+    def retrieve_ontology_term_labels_and_synonyms(self, term, property_name, convention, attribute_type: str):
         """Retrieve an individual term label from an ontology
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 2]
```diff
@@ -495,9 +506,10 @@ def insert_array_ontology_label_row_data(
         for id in row[property_name]:
             label_lookup = ""
             try:
-                label_lookup = retriever.retrieve_ontology_term_label(
+                labels_and_synonyms = retriever.retrieve_ontology_term_labels_and_synonyms(
                     id, property_name, convention, "array"
                 )
+                label_lookup = labels_and_synonyms.get('label')
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 3]
```diff
@@ -538,9 +550,10 @@ def insert_ontology_label_row_data(
         # for optional columns, try to fill it in
         property_type = convention["properties"][property_name]["type"]
         try:
-            label = retriever.retrieve_ontology_term_label(
+            labels_and_synonyms = retriever.retrieve_ontology_term_labels_and_synonyms(
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 4]
```diff
@@ -1032,6 +1044,23 @@ def exit_if_errors(metadata):
         exit(1)
     return errors
 
+def is_label_or_synonym(labels, provided_label):
+    """Return the best possible match from a list of candidates
+    :param labels: cached ontology labels from retriever.retrieve_ontology_term_label
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 5]
```diff
@@ -1059,11 +1088,13 @@ def validate_collected_ontology_data(metadata, convention):
 
             try:
                 attribute_type = convention["properties"][property_name]["type"]
-                matched_label_for_id = retriever.retrieve_ontology_term_label(
+                # get actual label along with synonyms for more robust matching
+                labels_and_synonyms = retriever.retrieve_ontology_term_labels_and_synonyms(
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 6]
```diff
@@ -1059,11 +1088,13 @@ def validate_collected_ontology_data(metadata, convention):
...
                     ontology_id, property_name, convention, attribute_type
                 )
-
-                if matched_label_for_id != ontology_label:
+                is_match = is_label_or_synonym(labels_and_synonyms, ontology_label)
+                if not is_match:
+                    matched_label_for_id = labels_and_synonyms.get("label")
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 7]
```diff
@@ -538,9 +550,10 @@ def insert_ontology_label_row_data(
         # for optional columns, try to fill it in
         property_type = convention["properties"][property_name]["type"]
         try:
-            label = retriever.retrieve_ontology_term_label(
+            labels_and_synonyms = retriever.retrieve_ontology_term_labels_and_synonyms(
                 id, property_name, convention, property_type
             )
+            label = labels_and_synonyms.get('label')
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

File: ingest/validation/validate_metadata.py [hunk 8]
```diff
@@ -495,9 +506,10 @@ def insert_array_ontology_label_row_data(
         for id in row[property_name]:
             label_lookup = ""
             try:
-                label_lookup = retriever.retrieve_ontology_term_label(
+                labels_and_synonyms = retriever.retrieve_ontology_term_labels_and_synonyms(
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: 94e8cae

**Review by eweitz — APPROVED** [review]

**Review by jlchang — APPROVED** [review]
This is great, Jon. I'm so glad that we can allow synonyms in our validation. I'll find the right place in our documentation to indicate the newly allowed exception and also make a ticket to find the right place in the Alexandria documentation and send over a PR to update their docs so end users know they should operate on the IDs rather than the labels because labels can evolve over time.

**Pull request merged**